File: Python3/601 - 800/743 Network Delay Time.py
# 104ms, 97.70%
class Solution:
    def networkDelayTime(self, times, N, K):
        from collections import defaultdict
        nodes = defaultdict(dict)
        Q = set(range(N))
        for u, v, w in times:
            nodes[u - 1][v - 1] = w
        dist = [float('inf')] * N
        dist[K - 1] = 0
        while Q:
            u = None
            for node in Q:
                if u is None or dist[node] < dist[u]: u = node
            Q.remove(u)
            for v in nodes[u]:
                dist[v] = min(dist[v], dist[u] + nodes[u][v])
        d = max(dist)
        return -1 if d == float('inf') else d
    

File: 601 - 800/743. Network Delay Time.cc
// 90ms, 48.68ms
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int networkDelayTime(vector<vector<int>> &times, int N, int K) {
        int *dist = new int[N + 1];
        fill(dist, dist + N + 1, INT_MAX);
        dist[K] = 0;

        for (int i = 0; i < N; ++i)
            for (auto &t : times) {
                int u = t[0], v = t[1], w = t[2];
                if (dist[u] != INT_MAX && dist[v] > dist[u] + w) dist[v] = dist[u] + w;
            }
        int maxwait = 0;
        for (int i = 1; i <= N; ++i)
            if (maxwait < dist[i]) maxwait = dist[i];

        return maxwait == INT_MAX ? -1 : maxwait;
    }
};
